### PR TITLE
feature(builder): rewrite builder in Go

### PR DESCRIPTION
**Pull request by technosophos**

### CODE REVIEW

**Comment by mboersma** [comment]
Needs rebase.

File: builder/README.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 # Deis Builder
 
-A Docker image that builds Docker images, for use in the [Deis](http://deis.io/) open source PaaS.
+**WARNING:** This is the experimental Go-based builder.
```
Comment (carmstrong) [inline builder/README.md]: Should we remove this? After the merge, this will indeed be the official builder :)

**Comment by mboersma** [comment]
This has some "[WIP]" commit messages, but when that's cleaned up this is ready to merge. :sparkles:

**Comment by technosophos** [comment]
I am going to rebase, squash commits, and then get ready to merge.

File: builder/confd/confd.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+//
+// Params:
+// - node (string): The etcd node to use. (Only etcd is currently supported)
+//
+// Returns:
+// - The []bytes from stdout and stderr when running the program.
+//
+func RunOnce(c cookoo.Context, p *cookoo.Params) (interface{}, cookoo.Interrupt) {
+
+	node := p.Get("node", "127.0.0.1:4001").(string)
```
Comment (kalbasit) [inline builder/confd/confd.go]: can we move this URL to a constant or a var for easy maintenance?

File: builder/confd/confd.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+	dargs := []string{"-onetime", "-node", node, "-log-level", "error"}
+
+	fmt.Println("Building confd templates. This may take a moment.")
```
Comment (kalbasit) [inline builder/confd/confd.go]: Perhaps `log.Println` to get the timestamp as well?

File: builder/confd/confd.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+			log.Infof(c, "Templates generated for %s on run %d", node, i)
+			return out, nil
+		} else {
+			log.Debugf(c, "Recoverable error: %s", err)
+			log.Debugf(c, "Output: %q", out)
+			lasterr = err
+		}
+
+		time.Sleep(timeout)
+		log.Infof(c, "Re-trying template build. (Elapsed time: %d)", i*3)
```
Comment (kalbasit) [inline builder/confd/confd.go]: `i*3` is confusing you probably mean `i*timeout`? Also on the first run `i` is 0 but 3 seconds has actually elapsed so maybe `(i+1)*timeout`

File: builder/confd/confd.go (hunk 2 above)
Comment (technosophos) [inline builder/confd/confd.go]: That's a default.

File: builder/confd/confd.go (hunk 3 above)
Comment (technosophos) [inline builder/confd/confd.go]: Good catch. I'll fix that.

File: builder/confd/confd.go (hunk 4 above)
Comment (technosophos) [inline builder/confd/confd.go]: That was just for debugging, but I will fix it and leave it in there. It might be useful to know.

**Comment by technosophos** [comment]
Closing this pull request because of the bad `rebase` that caught two extra commits. I'll cherry-pick and open a new PR.

Thanks @kalbasit for the code review. I think I incorporated all of your recommendations in the latest version.

**Pull request closed without merging**